core/dns.go

```diff
@@ -79,41 +108,47 @@ func (dnsResolver *DNSResolverImpl) LookupTXT(hostname string) ([]string, time.D
 	return txt, rtt, err
 }
 
+func isPrivate(ip net.IP) bool {
+	return privateNetworkA.Contains(ip) || privateNetworkB.Contains(ip) || privateNetworkC.Contains(ip)
+}
+
 // LookupHost sends a DNS query to find all A/AAAA records associated with
 // the provided hostname.
-func (dnsResolver *DNSResolverImpl) LookupHost(hostname string) ([]net.IP, time.Duration, time.Duration, error) {
+func (dnsResolver *DNSResolverImpl) LookupHost(hostname string, filter AddrFilter) ([]net.IP, time.Duration, time.Duration, error) {
 	var addrs []net.IP
 	var answers []dns.RR
+	var aRtt time.Duration
+	var aaaaRtt time.Duration
 
 	r, aRtt, err := dnsResolver.ExchangeOne(hostname, dns.TypeA)
 	if err != nil {
-		return addrs, 0, 0, err
+		return addrs, aRtt, 0, err
 	}
 	if r.Rcode != dns.RcodeSuccess {
 		err = fmt.Errorf("DNS failure: %d-%s for A query", r.Rcode, dns.RcodeToString[r.Rcode])
 		return nil, aRtt, 0, err
 	}
-
 	answers = append(answers, r.Answer...)
 
-	r, aaaaRtt, err := dnsResolver.ExchangeOne(hostname, dns.TypeAAAA)
-	if err != nil {
-		return addrs, aRtt, 0, err
-	}
-	if r.Rcode != dns.RcodeSuccess {
-		err = fmt.Errorf("DNS failure: %d-%s for AAAA query", r.Rcode, dns.RcodeToString[r.Rcode])
-		return nil, aRtt, aaaaRtt, err
+	if filter != ipv4OnlyFilter {
+		r, aaaaRtt, err := dnsResolver.ExchangeOne(hostname, dns.TypeAAAA)
+		if err != nil {
+			return addrs, aRtt, aaaaRtt, err
+		}
+		if r.Rcode != dns.RcodeSuccess {
+			err = fmt.Errorf("DNS failure: %d-%s for AAAA query", r.Rcode, dns.RcodeToString[r.Rcode])
+			return nil, aRtt, aaaaRtt, err
+		}
+		answers = append(answers, r.Answer...)
 	}
 
-	answers = append(answers, r.Answer...)
-
 	for _, answer := range answers {
 		if answer.Header().Rrtype == dns.TypeA {
-			if a, ok := answer.(*dns.A); ok {
+			if a, ok := answer.(*dns.A); ok && a.A.To4() != nil && !isPrivate(a.A) {
 				addrs = append(addrs, a.A)
 			}
 		} else if answer.Header().Rrtype == dns.TypeAAAA {
-			if aaaa, ok := answer.(*dns.AAAA); ok {
+			if aaaa, ok := answer.(*dns.AAAA); ok && aaaa.AAAA.To16() != nil && !isPrivate(aaaa.AAAA) && filter != ipv4OnlyFilter {
```

Comment:
This is to check that the bytes `miekg/dns` unmarshals into `net.IP` actually correspond to a IPv4/6 address since the struct can represent both. In this case a server could respond with a AAAA record but have the bytes packed so that when unmarshaled it is actually a IPv4 address.